### ISSUE TITLE
Core300s - Polling Interval

### ISSUE THREAD

**Issue by martynjsimpson** [dag81/vesyncOpenhabPyPort#2]
You mentioned that you might be looking for support in testing with a Core300s. I have one and the core functionality appears to be working fine with the exception of the Polling.

The Polling interval is set to 60 seconds. The "Polling interval for Air Filters" is set to 5 seconds.

Watching the events.log shows that no polling is taking place UNLESS you issue any command to the device. For example, if you send the "Display Off" command, the value for AirQualiyPPM25 updates. If you touch nothing, the values does not update.

My device name is "WindMachine" (internal joke with my wife - don't ask).

```
root@automate:/home/system# tail -f /var/log/openhab/events.log | grep WindMachine_
2022-01-28 11:17:28.695 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'WindMachine_Display' received command OFF
2022-01-28 11:17:28.700 [INFO ] [penhab.event.ItemStatePredictedEvent] - Item 'WindMachine_Display' predicted to become OFF
2022-01-28 11:17:28.702 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Display' changed from ON to OFF
2022-01-28 11:17:30.268 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 612 to 628
```

**Comment by dag81** [comment]
Investigating - did you use the jar file from the main branch, also with which version of OpenHab?

**Comment by martynjsimpson** [comment]
OpenHab 3.2.0 Release Build.

Yea I took the prebuilt jar from main - org.openhab.binding.vesync-3.2.0-SNAPSHOT.jar.

I don't have any other "add ons" (outside the bindings provided by OpenHab.

**Comment by dag81** [comment]
Hi @martynjsimpson,

I installed 3.2.0 Stable Build which I would assume is the same as yours.

I added the jar, then added the VeSync Bridge and configured. Then added one of the Core400S Air Filters here (not named as exciting as Wind_Machine unfortunately). The polling engine is the same.

I then linked the operation mode to a new point with the standard settings.

In the log when the mode changed I got:

21:42:46.768 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'LoungeAirPurifier_Operationmode' changed from NULL to manual
21:43:18.260 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'LoungeAirPurifier_Operationmode' changed from manual to auto
21:45:47.810 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'LoungeAirPurifier_Operationmode' changed from auto to manual
21:45:58.454 [INFO ] [openhab.event.ItemStateChangedEvent  ] - Item 'LoungeAirPurifier_Operationmode' changed from manual to auto

which was within 5 seconds of each physical interaction with the device. To see the actual polling happening then you can set in the console:

log:set DEBUG org.openhab.binding.vesync

You should see every 5 seconds a poll request and response happening 🤞 Please confirm and I can close, unless there is something else, or another flow I'm missing / some errors output to the console please?

I've merged the current build, so the latest and greatest build is on main.

**Comment by martynjsimpson** [comment]
I think i managed to pull the error. 

See below.

```
2022-01-29 12:49:52.682 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.NullPointerException: null
	at org.openhab.binding.vesync.internal.handlers.VeSyncDeviceAirPurifierHandler.processV2BypassPoll(VeSyncDeviceAirPurifierHandler.java:381) ~[?:?]
	at org.openhab.binding.vesync.internal.handlers.VeSyncDeviceAirPurifierHandler.pollForDeviceData(VeSyncDeviceAirPurifierHandler.java:259) ~[?:?]
	at org.openhab.binding.vesync.internal.handlers.VeSyncBaseDeviceHandler.pollForUpdate(VeSyncBaseDeviceHandler.java:335) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]

```

**Comment by martynjsimpson** [comment]
Here are some more potentially useful logs.

I can't figure out if the polling is happening or not but;

It looks like the value returned for airQuality is "-1" which may be throwing the error

```
2022-01-29 13:10:29.162 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - POST @ https://smartapi.vesync.com/cloud/v1/deviceManaged/devices with content
{
  "pageNo": "1",
  "pageSize": "100",
  "accountID": "6258351",
  "token": "oppD5LVncz45N7g_Vwk0CPY53apMyTfoM600KFqC94tAMSR1zQ==",
  "timeZone": "America/New_York",
  "acceptLanguage": "en",
  "appVersion": "2.5.1",
  "phoneBrand": "SM N9005",
  "phoneOS": "Android",
  "traceId": "1643461829162",
  "method": "devices"
}
2022-01-29 13:10:29.583 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - Got OK response {"traceId":"1643461829162","code":0,"msg":"request success","result":{"total":1,"pageSize":100,"pageNo":1,"list":[{"deviceRegion":"EU","isOwner":true,"authKey":null,"deviceName":"Wind Machine","deviceImg":"https://image.vesync.com/defaultImages/Core_300S_Series/icon_300_purifier_160.png","cid":"vsaq9a9767fc48ed998aa498d4858e68","deviceStatus":"on","connectionStatus":"online","connectionType":"WiFi+BTOnboarding+BTNotify","deviceType":"Core300S","type":"wifi-air","uuid":"46897852-e77b-45b5-a10a-740f73ad6089","configModule":"WiFiBTOnboardingNotify_AirPurifier_Core300S_EU","macID":"a8:48:fa:32:be:ba","mode":null,"speed":null,"extension":{"airQuality":-1,"airQualityLevel":4,"mode":"auto","fanSpeedLevel":"2"},"currentFirmVersion":null,"subDeviceNo":null,"subDeviceType":null,"deviceFirstSetupTime":"Jan 27, 2022 3:38:43 PM","deviceProp":null}]}}
2022-01-29 13:10:29.583 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - Found device : Wind Machine, type: wifi-air, deviceType: Core300S, connectionState: online, deviceStatus: on, deviceRegion: EU, cid: vsaq9a9767fc48ed998aa498d4858e68, configModule: WiFiBTOnboardingNotify_AirPurifier_Core300S_EU, macID: a8:48:fa:32:be:ba, uuidL 46897852-e77b-45b5-a10a-740f73ad6089
2022-01-29 13:10:29.584 [DEBUG] [nal.handlers.VeSyncBaseDeviceHandler] - Searching for device mac id : a8:48:fa:32:be:ba

```

**Comment by martynjsimpson** [comment]
Oh and for reference I update the JAR to the latest build from last night.

**Comment by dag81** [comment]
Yeah defo a bug, I've adding some logging and wrappers around where I think the data must be missing. A test build is here to validate the bug and get info from:

https://github.com/dag81/vesyncOpenhabPyPort/tree/main/bug%20validation%20builds/Core300S

Hopefully it that works it proves where the error is to a degree. If in the logged output you can spot this test - I need the big json data afterwards - so I can add test cases and clean up properly thanks. I can confirm its what I think from the json after this message.

ERROR CHECKING: Please send payload for ERROR CHECKING : 

Hopefully that's okay thanks in advance.

**Comment by martynjsimpson** [comment]
Loaded the build as provided.

Here is the logs you are after. I think.

EDIT Formatting.

```
2022-01-29 13:24:54.370 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - POST @ https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2 with content
{
  "deviceRegion": "EU",
  "debugMode": false,
  "cid": "<Removed>",
  "configModule": "WiFiBTOnboardingNotify_AirPurifier_Core300S_EU",
  "payload": {
    "method": "getPurifierStatus",
    "source": "APP",
    "data": null
  },
  "accountID": "<Removed>",
  "token": "<Removed>",
  "timeZone": "America/New_York",
  "acceptLanguage": "en",
  "appVersion": "2.5.1",
  "phoneBrand": "SM N9005",
  "phoneOS": "Android",
  "traceId": "1643462694369",
  "method": "bypassV2"
}
2022-01-29 13:24:54.576 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - Got OK response {"traceId":"1643462694369","code":0,"msg":"request success","result":{"traceId":"1643462694369","code":0,"result":{"enabled":true,"filter_life":99,"mode":"auto","level":1,"air_quality":2,"air_quality_value":72,"display":true,"child_lock":false,"configuration":{"display":true,"display_forever":true,"auto_preference":{"type":"quiet","room_size":0}},"device_error_code":0}}}
2022-01-29 13:24:54.578 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - ERROR CHECKING: Please send payload for ERROR CHECKING : {"traceId":"1643462694369","code":0,"msg":"request success","result":{"traceId":"1643462694369","code":0,"result":{"enabled":true,"filter_life":99,"mode":"auto","level":1,"air_quality":2,"air_quality_value":72,"display":true,"child_lock":false,"configuration":{"display":true,"display_forever":true,"auto_preference":{"type":"quiet","room_size":0}},"device_error_code":0}}}
2022-01-29 13:24:54.583 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - Extension is null: true
2022-01-29 13:24:59.588 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - POST @ https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2 with content
{
  "deviceRegion": "EU",
  "debugMode": false,
  "cid": "<Removed>",
  "configModule": "WiFiBTOnboardingNotify_AirPurifier_Core300S_EU",
  "payload": {
    "method": "getPurifierStatus",
    "source": "APP",
    "data": null
  },
  "accountID": "<Removed>",
  "token": "<Removed>",
  "timeZone": "America/New_York",
  "acceptLanguage": "en",
  "appVersion": "2.5.1",
  "phoneBrand": "SM N9005",
  "phoneOS": "Android",
  "traceId": "1643462699588",
  "method": "bypassV2"
}
2022-01-29 13:24:59.890 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - Got OK response {"traceId":"1643462699588","code":0,"msg":"request success","result":{"traceId":"1643462699588","code":0,"result":{"enabled":true,"filter_life":99,"mode":"auto","level":1,"air_quality":3,"air_quality_value":82,"display":true,"child_lock":false,"configuration":{"display":true,"display_forever":true,"auto_preference":{"type":"quiet","room_size":0}},"device_error_code":0}}}
2022-01-29 13:24:59.891 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - ERROR CHECKING: Please send payload for ERROR CHECKING : {"traceId":"1643462699588","code":0,"msg":"request success","result":{"traceId":"1643462699588","code":0,"result":{"enabled":true,"filter_life":99,"mode":"auto","level":1,"air_quality":3,"air_quality_value":82,"display":true,"child_lock":false,"configuration":{"display":true,"display_forever":true,"auto_preference":{"type":"quiet","room_size":0}},"device_error_code":0}}}
2022-01-29 13:24:59.892 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - Extension is null: true
```

**Comment by dag81** [comment]
Thats great thank you... just to confirm one final thing. Im trying to figure out if the Core300S doesn't have that at all, or if it's because the payloads missing when that feature hasn't been used yet. Can you on the app set a timer for the unit to switch off - If you can does the error go away? I just want to see if its just excluded until used, or always missing for the 300S. Thanks again.

**Comment by martynjsimpson** [comment]
From within the App I can see (but have not tested);

- Timers
- Schedule
- PM2.5 Value (air quality)
- Fan Speed (Low, Med, High, Sleep)
- On/ Off
- Filter Life 
- Cumulative Running hours
- Cumulative purified

Per your request.

From within the app I setup a 5 min timer.

Here are the logs.

```
2022-01-29 13:35:21.762 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - Extension is null: true
2022-01-29 13:35:26.763 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - POST @ https://smartapi.vesync.com/cloud/v2/deviceManaged/bypassV2 with content
{
  "deviceRegion": "EU",
  "debugMode": false,
  "cid": "<Removed>",
  "configModule": "WiFiBTOnboardingNotify_AirPurifier_Core300S_EU",
  "payload": {
    "method": "getPurifierStatus",
    "source": "APP",
    "data": null
  },
  "accountID": "<Removed>",
  "token": "<Removed>,
  "timeZone": "America/New_York",
  "acceptLanguage": "en",
  "appVersion": "2.5.1",
  "phoneBrand": "SM N9005",
  "phoneOS": "Android",
  "traceId": "1643463326763",
  "method": "bypassV2"
}
2022-01-29 13:35:27.085 [DEBUG] [esync.internal.api.VesyncV2ApiHelper] - Got OK response {"traceId":"1643463326763","code":0,"msg":"request success","result":{"traceId":"1643463326763","code":0,"result":{"enabled":true,"filter_life":99,"mode":"manual","level":2,"air_quality":4,"air_quality_value":926,"display":true,"child_lock":false,"configuration":{"display":true,"display_forever":true,"auto_preference":{"type":"quiet","room_size":0}},"device_error_code":0}}}
2022-01-29 13:35:27.087 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - ERROR CHECKING: Please send payload for ERROR CHECKING : {"traceId":"1643463326763","code":0,"msg":"request success","result":{"traceId":"1643463326763","code":0,"result":{"enabled":true,"filter_life":99,"mode":"manual","level":2,"air_quality":4,"air_quality_value":926,"display":true,"child_lock":false,"configuration":{"display":true,"display_forever":true,"auto_preference":{"type":"quiet","room_size":0}},"device_error_code":0}}}
2022-01-29 13:35:27.090 [WARN ] [dlers.VeSyncDeviceAirPurifierHandler] - Extension is null: true
```
I also manually issued a POST request to the endpoint while the time was running.

```
REQUEST:
{
  "deviceRegion": "EU",
  "debugMode": false,
  "cid": "<Removed>",
  "configModule": "WiFiBTOnboardingNotify_AirPurifier_Core300S_EU",
  "payload": {
    "method": "getPurifierStatus",
    "source": "APP",
    "data": null
  },
  "accountID": "<Removed>",
  "token": "<Removed>",
  "timeZone": "America/New_York",
  "acceptLanguage": "en",
  "appVersion": "2.5.1",
  "phoneBrand": "SM N9005",
  "phoneOS": "Android",
  "traceId": "1643463326763",
  "method": "bypassV2"
}


RESPONSE
{
    "traceId": "1643463326763",
    "code": 0,
    "msg": "request success",
    "result": {
        "traceId": "1643463326763",
        "code": 0,
        "result": {
            "enabled": true,
            "filter_life": 99,
            "mode": "manual",
            "level": 2,
            "air_quality": 4,
            "air_quality_value": 131,
            "display": true,
            "child_lock": false,
            "configuration": {
                "display": true,
                "display_forever": true,
                "auto_preference": {
                    "type": "quiet",
                    "room_size": 0
                }
            },
            "device_error_code": 0
        }
    }
}
```

**Comment by martynjsimpson** [comment]
And if it helps, after the timer timed out the following POST response was retrieved manually. (same request as above)

```
{
    "traceId": "1643463326763",
    "code": 0,
    "msg": "request success",
    "result": {
        "traceId": "1643463326763",
        "code": 0,
        "result": {
            "enabled": false,
            "filter_life": 99,
            "mode": "manual",
            "level": 2,
            "air_quality": 3,
            "air_quality_value": 85,
            "display": false,
            "child_lock": false,
            "configuration": {
                "display": true,
                "display_forever": true,
                "auto_preference": {
                    "type": "quiet",
                    "room_size": 0
                }
            },
            "device_error_code": 0
        }
    }
}
```

**Comment by dag81** [comment]
Many thanks @martynjsimpson, yeah okay so it looks like the 300S doesn't include some data the 400S has. i'll use those payloads to update, and remove the extension based bits so only the 400S uses them. Got to shoot out now but will hopefully be able to update main this evening. I'm assuming now the parsing isn't crashing out its polling with that manual build atm?

**Comment by martynjsimpson** [comment]
Heya,

Yes the polling is no longer failing and the values per below are updated every 5 seconds (inline with poling interval value).

Let me know if you need anything more. Enjoy your day.

```
2022-01-29 13:58:31.280 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 82 to 101
2022-01-29 13:58:36.492 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airquality' changed from 3 to 4
2022-01-29 13:58:36.492 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 101 to 117
2022-01-29 13:58:41.807 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 117 to 137
2022-01-29 13:58:47.118 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 137 to 129
2022-01-29 13:58:52.595 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 129 to 176
2022-01-29 13:58:57.974 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 176 to 192
2022-01-29 13:59:03.239 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 192 to 170
2022-01-29 13:59:08.648 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 170 to 139
2022-01-29 13:59:13.861 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airquality' changed from 4 to 3
2022-01-29 13:59:13.862 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 139 to 99
2022-01-29 13:59:19.173 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 99 to 107
2022-01-29 13:59:24.495 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'WindMachine_Airqualityppm25' changed from 107 to 103
```

**Comment by dag81** [comment]
Hi @martynjsimpson, Many thanks for testing with the 300S 👍 

I've updated the build to what I plan on merging to master, if you get a chance just to verify it works as expected, before I do - with the 300S that would be great. (I've checked with my 400S - but that doesn't mean there's something else I've missed.

The jar file is located here until I merge the bugfix in: 

https://github.com/dag81/vesyncOpenhabPyPort/blob/bugfix/2/manual%20builds/org.openhab.binding.vesync-3.2.0-SNAPSHOT.jar

Hopefully there's no big bits I missed when mapping it over otherwise.

**Comment by martynjsimpson** [comment]
Heya,

Looks all ok to me.

I removed all Items, removed everything from the model, removed the bindings, removed previous JAR (apparently renaming it to .xxx resulted In it still being loaded), installed the linked JAR, re-created the Binding, Thing and Items.

You should be good to merge this with main.

Cheers.

**Comment by dag81** [comment]
Core300S merged to main in commit: 3f992fbe9961624c09333967aef728e05dee4a02 - many thanks @martynjsimpson
